src/brevitas_examples/common/learned_round/learned_round_optimizer.py

```diff
@@ -508,25 +510,32 @@ def _optimize_learned_round_block(
         init_loss = -1.0
         last_best_iter = self.iters
 
+        scaler = None
+        use_amp = next(block.parameters()).dtype == torch.float32
```

Comment:
Why torch.float32? If the LLM dtype is torch.float16, this would be False and no gradient scaling would be used.

Reply:
Gradient scaling and AMP are to be used only with higher precision data type (float32) and take advantage of fp16 in selected layers.

If the network is already cast to float16, we don't need mixed precision, and we do manual scaling of the loss later

Reply:
GradScaler specifically raise an error if the original network is fully cast to float16